docs/reference/fields.txt

```diff
@@ -576,93 +578,341 @@ you have three options:
+      end
+
+      # Converts the object that was supplied to a criteria and converts it
+      # into a query-friendly form. Returns invalid values as is.
+      def evolve(object)
+        MAPPING.fetch(object, object)
+      end
+    end
+  end
+
+  class Profile
+    include Mongoid::Document
+    field :color, type: ColorMapping
+  end
+
+  profile = Profile.new(color: 'white')
+  profile.color
+  # => "white"
+
+  # Writes 0 to color field
+  profile.save!
+
+
+.. _custom-field-options:
+
+Custom Field Options
+====================
+
+You may define custom options for the ``field`` macro function
+which extend its behavior at the your time model classes are loaded.
+
+As an example, we will define a ``:required`` option which will add a presence
+validator for the field. First, declare the new field option in an initializer,
+specifiying its handler function as a block:
+
+.. code-block:: ruby
+
+  # in /config/initializers/mongoid_custom_fields.rb
+
+  Mongoid::Fields.option :required do |model, field, value|
+    model.validates_presence_of field if value
+  end
+
+Then, use it your model class:
+
+.. code-block:: ruby
+
+  class Person
+    include Mongoid::Document
+
+    field :name, type: String, required: true
+  end
+
+Note that the handler function will be invoked whenever the option is used
+in the field definition, even if the option's value is false or nil.
+
+
+Symbol/String Field Type Aliases
+================================
+
+Mongoid permits using symbols or strings instead of classes to specify the
+type of fields, for example:
+
+.. code-block:: ruby
+
+  class Order
+    include Mongoid::Document
+
+    field :state, type: :integer
+    # Equivalent to:
+    field :state, type: "integer"
+    # Equivalent to:
+    field :state, type: Integer
+  end
+
+Only standard field types as listed below can be specified using symbols or
+strings in this manner. Mongoid recognizes the following expansions:
+
+- ``:array`` => ``Array``
+- ``:big_decimal`` => ``BigDecimal``
+- ``:binary`` => ``BSON::Binary``
+- ``:boolean`` => ``Mongoid::Boolean``
+- ``:date`` => ``Date``
+- ``:date_time`` => ``DateTime``
+- ``:float`` => ``Float``
+- ``:hash`` => ``Hash``
+- ``:integer`` => ``Integer``
+- ``:object_id`` => ``BSON::ObjectId``
+- ``:range`` => ``Range``
+- ``:regexp`` => ``Regexp``
+- ``:set`` => ``Set``
+- ``:string`` => ``String``
+- ``:stringified_symbol`` => ``StringifiedSymbol``
+- ``:symbol`` => ``Symbol``
+- ``:time`` => ``Time``
+
+
+.. _custom-id:
+
+Custom IDs
+==========
+
+By default, Mongoid defines the ``_id`` field on documents to contain a
+``BSON::ObjectId`` value which is automatically generated by Mongoid.
+
+It is possible to replace the ``_id`` field definition to change the type
+of the ``_id`` values or have different default values:
+
+.. code-block:: ruby
+
+  class Band
+    include Mongoid::Document
+    field :name, type: String
+    field :_id, type: String, default: ->{ name }
```

Comment:
Do we want to note that associations are not available inside the default proc?

i.e. the following won't work:
```
class Band
     include Mongoid::Document
     field :name, type: String
     field :_id, type: String, default: ->{ label.name }
     
     embeds_one :label
end
```

Reply:
https://jira.mongodb.org/browse/MONGOID-3267 would imply to me that it is in fact possible?